spec/lang/well-formed.md

```diff
@@ -670,13 +670,15 @@ fn is_atomic_binop(op: IntBinOp) -> bool {
 impl Terminator {
     fn check_wf<T: Target>(
         self,
+        block: BasicBlock,
         func: Function,
         prog: Program,
     ) -> Result<()> {
         use Terminator::*;
         match self {
             Goto(block_name) => {
                 ensure_wf(func.blocks.contains_key(block_name), "Terminator::Goto: next block does not exist")?;
+                ensure_wf(block.kind == func.get_block_kind(block_name), "Terminator::Goto: next block has the wrong block kind")?;
```

Comment:
You are looking up the block twice here. Instead, please do something like:
```rust
let Some(next_block) = func.blocks.get(block_name) else { throw_ill_formed!("...") };
ensure_wf(next_block.kind == block_kind, "...");
```

This of course applies throughout this function.

Maybe you'll want to define a little helper function, `check_next_block`, locally in this file, to make this less repetetive.